### DISCUSSION TITLE
Type does not flow properly when using the `set` during creation

### DISCUSSION THREAD

**Discussion by jsgoupil** [pmndrs/zustand#2665] [Bug report] [answered]
### Bug Description

With the following code:

```
import { create } from 'zustand';

const initialState = { b: true};
const working = create(() =>initialState);
const notWorking = create(set => initialState);
const workaround = create<typeof initialState>(set => initialState);
```

The create with the `set` present will make the StoreApi have an `unknown` as the model.

![image](https://github.com/user-attachments/assets/b0567431-37d2-4723-9e46-6d7a47f36946)
![image](https://github.com/user-attachments/assets/2d1a10fa-d0fb-401f-b5cb-2a61ef63f92f)

The workaround only works if my `initialState` does not employ `set`.
Otherwise, I have to duplicate everything like:

```
const workaround2 = create<{ z: boolean }>(set => ({ z: true }));
```

Which becomes very cumbersome.

Even if this is the "norm", is there a better workaround?

TypeScript:
5.2.2

### Reproduction Link

pasted

**Comment by dbritto-dev** [comment]
> ### Bug Description
> With the following code:
> 
> ```
> import { create } from 'zustand';
> 
> const initialState = { b: true};
> const working = create(() =>initialState);
> const notWorking = create(set => initialState);
> const workaround = create<typeof initialState>(set => initialState);
> ```
> 
> The create with the `set` present will make the StoreApi have an `unknown` as the model.
> 
> ![image](https://private-user-images.githubusercontent.com/808649/355594883-b0567431-37d2-4723-9e46-6d7a47f36946.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5ODAzNTEsIm5iZiI6MTcyMjk4MDA1MSwicGF0aCI6Ii84MDg2NDkvMzU1NTk0ODgzLWIwNTY3NDMxLTM3ZDItNDcyMy05ZTQ2LTZkN2E0N2YzNjk0Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwNlQyMTM0MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZjFlNjgwN2UxZDgyNDMzMGM4NjI2N2NmYTQ4NDFkM2VhYzUwZTRkNmYxNWFmN2Q5ODQ2MWI1ZTkyYjE5YjRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-wxSe7JtTzSrU3VA18dkoRCmJYXRSuhO8fx0yelfBS8) ![image](https://private-user-images.githubusercontent.com/808649/355594955-2d1a10fa-d0fb-401f-b5cb-2a61ef63f92f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5ODAzNTEsIm5iZiI6MTcyMjk4MDA1MSwicGF0aCI6Ii84MDg2NDkvMzU1NTk0OTU1LTJkMWExMGZhLWQwZmItNDAxZi1iNWNiLTJhNjFlZjYzZjkyZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwNlQyMTM0MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYzg1NjM3MGYxZDBhMDNlZjE1MmI3YmJhZTU1Mzg4MGRmYjkyNDFlZTcwY2MyM2EyYmZhYjYxYTYyMzhjMzlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.r8BUC09nJZPz18q69-1y6dJTpphZAUJlDl3FH_xcNnw)
> 
> The workaround only works if my `initialState` does not employ `set`. Otherwise, I have to duplicate everything like:
> 
> ```
> const workaround2 = create<{ z: boolean }>(set => ({ z: true }));
> ```
> 
> Which becomes very cumbersome.
> 
> Even if this is the "norm", is there a better workaround?
> 
> TypeScript: 5.2.2
> 
> ### Reproduction Link
> pasted

JS middleware and TS -> `create()(...)`
JS no middleware or combine -> `create(...)`

**Reply by jsgoupil** [reply]
This doesn't work. It results to the same problem.

**Comment by dbritto-dev** [comment, marked as answer]
BTW, you don't need to explicitly type your store, you can use `combine` middleware to get fully inferred types ->

```ts
create(
  combine({ count: 0 }, (set) => ({
    increment: () => {
      set((state) => ({ count: state.count + 1 }));
    },
  })),
);
```

**Reply by jsgoupil** [reply]
I guess that's an option that I could look into. Thanks.

**Reply by dbritto-dev** [reply]
@jsgoupil on zustand you need to explicitly type your store if you are not using `combine` middleware
